src/library_manager/lib_manager.c

```diff
@@ -604,36 +588,57 @@ int lib_manager_load_library(uint32_t dma_id, uint32_t lib_id)
 				       &addr_align);
 #endif
 	if (ret < 0)
-		goto cleanup;
+		goto e_init;
 
 	/* allocate temporary manifest buffer */
 	man_tmp_buffer = (__sparse_force void __sparse_cache *)rballoc_align(0, SOF_MEM_CAPS_DMA,
 								MAN_MAX_SIZE_V1_8, addr_align);
 	if (!man_tmp_buffer) {
 		ret = -ENOMEM;
-		goto cleanup;
+		goto e_init;
 	}
 
 	ret = lib_manager_dma_buffer_alloc(&dma_ext, MAN_MAX_SIZE_V1_8, addr_align);
 	if (ret < 0)
-		goto cleanup;
+		goto e_buf;
 
 	ret = core_kcps_adjust(cpu_get_id(), CLK_MAX_CPU_HZ / 1000);
 	if (ret < 0)
 		goto cleanup;
 
-	/* Load manifest to temporary buffer */
+	dma_block_cfg.dest_address = dma_ext.dma_addr;
+
+	ret = dma_config(dma_ext.chan->dma->z_dev, dma_ext.chan->index, &config);
+	if (ret < 0)
+		goto cleanup;
+
+	ret = dma_start(dma_ext.chan->dma->z_dev, dma_ext.chan->index);
+	if (ret < 0)
+		goto cleanup;
+
+	/* Load manifest to temporary buffer. */
 	ret = lib_manager_store_data(&dma_ext, man_tmp_buffer, MAN_MAX_SIZE_V1_8);
 	if (ret < 0)
 		goto cleanup;
 
 	ret = lib_manager_store_library(&dma_ext, man_tmp_buffer, lib_id, addr_align);
+	if (ret < 0)
+		tr_err(&lib_manager_tr, "library loading error %d", ret);
+
+	ret2 = dma_stop(dma_ext.chan->dma->z_dev, dma_ext.chan->index);
+	if (ret2 < 0) {
+		tr_err(&lib_manager_tr, "error %d stopping DMA", ret);
+		if (!ret)
+			ret = ret2;
+	}
 
 cleanup:
 	core_kcps_adjust(cpu_get_id(), -(CLK_MAX_CPU_HZ / 1000));
 	rfree((void *)dma_ext.dma_addr);
-	rfree((__sparse_force void *)man_tmp_buffer);
 
+e_buf:
+	rfree((__sparse_force void *)man_tmp_buffer);
+e_init:
```

Comment:
I can't follow (maybe because of the diff format) how the transfers were started multiple times and why, which makes the change even less self-explanatory.

Reply:
Before this patch the DMA got re-configured for each transfer block which must be avoided to be able to follow the programming sequence of the DMA.
With this change we will configure and start the DMA once, receive all the chunks and then stop the DMA.